#VRML_SIM R2025a utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/floors/protos/Floor.proto"

WorldInfo {
  info [
    "An example of hot to solve the Inverted Pendulum problem using a PID controller"
  ]
  title "Inverted Pendulum"
  contactProperties [
    ContactProperties {
      material1 "robot_basis"
      material2 "floor"
      coulombFriction [
        0.2
      ]
    }
  ]
}
Viewpoint {
  orientation -0.07569796563581566 -0.10723057262243957 0.9913482850611397 4.832213449087991
  position -0.12861867122135423 2.0394164342182437 0.5574602111326571
  follow "robot:solid"
}
TexturedBackground {
}
Floor {
  size 1000 2
  appearance PBRAppearance {
    baseColorMap ImageTexture {
      url [
        "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/default/worlds/textures/checkered_marble.jpg"
      ]
    }
    roughness 1
    metalness 0
  }
}
Robot {
  translation -1.2510737267367886e-06 3.3198404211077093e-12 -0.006092006374108783
  rotation 1.451107374603282e-10 -3.621635791696412e-16 1 3.14159
  children [
    SliderJoint {
      jointParameters JointParameters {
        position 0.04295101927329199
        axis 1 0 0
        dampingConstant 1.5
      }
      device [
        LinearMotor {
          name "horizontal_motor"
          maxForce 40
        }
        PositionSensor {
          name "horizontal position sensor"
        }
      ]
      endPoint Solid {
        translation 0.04295101927329199 0 0.06
        children [
          DEF ROBOT_SHAPE Shape {
            appearance PBRAppearance {
              baseColor 0.2443427176317998 0.704051270313573 0.1756923781185626
              roughness 1
              metalness 0
            }
            geometry Box {
              size 0.3 0.1 0.1
            }
          }
          DEF HIP HingeJoint {
            jointParameters HingeJointParameters {
              position -0.042944347001123095
              axis 0 1 0
              anchor 0 0 0.03
            }
            device [
              PositionSensor {
                name "hip"
              }
            ]
            endPoint DEF THIGH_BB Solid {
              translation -0.012927720320709043 -0.06099999999999998 0.3297213273147397
              rotation -6.598918352799927e-16 -1 -4.210086038632708e-15 0.04310574900000116
              children [
                Shape {
                  appearance PBRAppearance {
                    baseColor 0.8496833752956435 0.07072556649118791 0.09393453879606317
                    roughness 1
                    metalness 0
                  }
                  geometry DEF THIGH_BOX Box {
                    size 0.05 0.02 0.6
                  }
                }
              ]
              boundingObject USE THIGH_BOX
              physics Physics {
                density -1
                mass 0.05
                centerOfMass [
                  0 0.061 -0.27
                ]
              }
              linearVelocity -0.6149182958517772 1.6317409135319209e-06 -0.0006378095999451003
              angularVelocity 1.5855782193992987e-06 0.5975219763593167 1.378436651453086e-16
            }
          }
          PointLight {
            attenuation 0 0 1
            intensity 5
            location 0 0 2
          }
        ]
        contactMaterial "robot_basis"
        boundingObject USE ROBOT_SHAPE
        physics Physics {
          density -1
          mass 1
        }
        linearVelocity -0.6328597051464013 1.6793500542760763e-06 2.5873209574455077e-07
        angularVelocity 3.39730166712228e-11 1.2802040012475359e-05 -2.467474586934959e-17
      }
    }
  ]
  boundingObject Box {
    size 200 0.1 0.01
  }
  physics Physics {
    density -1
    mass 30
  }
  controller "my_Controller"
  linearVelocity 4.87328695617726e-05 -1.29317045729609e-10 1.9447336089797306e-07
  angularVelocity 7.749708333254139e-14 2.9227158296836087e-08 4.800587401938876e-20
}
